packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1449,10 +1458,37 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         }
     }
 
+    // Track how many times the container tries to replay its pending states.
+    // This happens when the connection state is changed.
+    // We reset the internal counter when we are able to process a local op.
+    // If this counter reaches a max, it's a good indicator that the container
+    // is not making progress and it is stuck in a retry loop.
+    private arePendingStatesDrained(): boolean {
```

Comment:
Name does not tell me much :) I'd also (based on name) not expect it to mutate state.
Not sure if it's easy to make it more readable, but maybe move closing container into here, and rename it to something similar to canSendOps (i.e. "can continue connection").